### ISSUE TITLE
add the capability to add padding for the hash traqnsformations (md5, ...)

### ISSUE THREAD

**Issue by terefang** [abhimanyu003/sttr#11]
an overview of standardized paddings is here https://en.wikipedia.org/wiki/Padding_%28cryptography%29

i saw the for 0.2 zero-padding has been merged

this enhancement request would rely upon that capability

**Comment by abhimanyu003** [comment]
Hello, @terefang thanks for the suggestion.
Can you please provide/share few good examples for MD5, SHA padding?

**Comment by terefang** [comment]
### ISO/IEC 7816-4:2005 Padding

from 1 to BS bytes are added as padding, the first byte is set to `0x80` all other bytes are zero

so for the byte sequence "s3cr3tPa55w0rd" (14 bytes)

* MD5 adds the bytes `0x80 0x00`
* SHA1 adds the bytes `0x80 0x00 0x00 0x00 0x00 0x00`

padding bytes will always be added

* the least padding is the byte `0x80`
* the longest padding is BS bytes all zero except for the first byte set to `0x80`

**Comment by ItsLychee** [comment]
@terefang Apologies for my asking, as I'm quite new to all of this stuff and I would like to help to the best of my ability, but is `s3cr3tPa55w0rd` the message digest in the examples you provided?

**Comment by rverst** [comment]
> Unless I'm gravely mistaken, it seems if Go's `crypto` module already does this for the user.

@ItsLychee yes, of course, the crypto packages of go are implemented as required by the individual algorithms. Otherwise, IMHO, they would be pretty useless.

@terefang, I honestly don't understand exactly what you're getting at. The noted zero-padding was an example implementation of mine to give an example of the new processor design. It simply prepends a fixed number of zeros to a number. Actually, the implementation is not even good, it would be better to specify the desired "width" of the number and then prepend a corresponding number of zeros. However, I haven't had time to tackle this yet.
